evm/src/executor/fork/backend.rs

```diff
@@ -335,23 +335,25 @@ where
         if let Poll::Ready(Some(ack)) = Pin::new(&mut pin.shutdown).poll_next(cx) {
```

Comment:
when i was debugging this, i added id (& counter incremented in `clone()`) to `SharedBackend`, so it was easy (for my usecase; when id == 0) to hack ideal behaviour & call `shutdown` only from last backend & proposed solution worked.

if every backend calls `shutdown`, it's still possible to hang when two backends are shutting down simultaneously (same mechanism as before + another possiblility because of "race" on `incoming.is_done()`). i didn't anticipated this either.

with spawning i can provoke yet another hang (old problem, from the beginning of this code), that i previously blamed on hyper (there is suitable mysterious issue in tracker & visible part of the problem is there), but now after experiencing how unusual is to write this, maybe it's time (for me) to think how this code works as a whole - now i'm paranoid about every state that leads to Pending.

Reply:
this is obviously beyond my current level of async experience, but i recognize that my code pushes evm beyond what is required for foundry to work.

Reply:
could use an atomic counter but still need a way to poll, and a channel seemed to be an easy solution at first.

> if every backend calls shutdown, it's still possible to hang when two backends are shutting down simultaneously

possibly, if the inner sender is dropped right after we received the second last sender's shutdown, but this shouldn't be a problem as we ignore errors on drop.

the whole problem was introduced when the cache was added which requires access to the db data, and backendhandler seemed to be the best fit for this.

Reply:
it looks like it somehow works, if shutdown's `if let Poll` is moved inside the loop & `if-break` is placed inside `if let Poll {} else { HERE }`, but i would like to test it more this time :).

Reply:
> possibly, if the inner sender is dropped right after we received the second last sender's shutdown, but this shouldn't be a problem as we ignore errors on drop.

"race" with waiting on ack was probably just random place where it sometimes printed before hang, but there is real problem with two shutdowns in queue ready to be polled & it slips to Pending between them without wake like before.